packages/platform-test-suite/test/functional/dapi/subscribeToBlockHeadersWithChainLocksHandlerFactory.spec.js

```diff
@@ -0,0 +1,178 @@
+        });
+      }
+
+      const rawChainLock = data.getChainLock();
+
+      if (rawChainLock) {
+        bestChainLock = new ChainLock(Buffer.from(rawChainLock));
+      }
+    });
+
+    let streamEnded = false;
+
+    stream.on('end', () => {
+      streamEnded = true;
+    });
+
+    let streamError;
+    stream.on('error', (e) => {
+      streamError = e;
+    });
+
+    while (!streamEnded) {
+      if (streamError) {
+        throw streamError;
+      }
+      await wait(1000);
+    }
+
+    expect(streamEnded).to.be.true();
+
+    // TODO: fetching blocks one by one takes too long. Implement getBlockHeaders in dapi-client
+    const fetchedBlocks = await Promise.all(
+      Array.from({ length: headersAmount })
+        .map(async (_, index) => new Block(await dapiClient.core.getBlockByHeight(index + 1))),
+    );
+
+    expect(historicalBlockHeaders.map((header) => header.hash))
+      .to.deep.equal(fetchedBlocks.map((block) => block.header.hash));
+    expect(bestChainLock.height).to.exist();
+  });
+
+  it('should respond with both new and historical data', async () => {
+    const blocksToGenerate = 5;
+    let latestChainLock = null;
+
+    const numHistoricalBlocks = 10;
+    const blockHeadersHashesFromStream = [];
+    const blockHeadersHashesGenerated = [];
+
+    let allHeadersSettled = false;
+
+    // Connect to the stream
+    const stream = await dapiClient.core.subscribeToBlockHeadersWithChainLocks(
+      {
+        fromBlockHeight: bestBlockHeight - numHistoricalBlocks + 1,
+      },
+    );
+
+    let streamEnded = false;
+    stream.on('data', (data) => {
+      const blockHeaders = data.getBlockHeaders();
+
+      if (blockHeaders) {
+        const list = blockHeaders.getHeadersList();
+        list.forEach((headerBytes) => {
+          const header = new BlockHeader(Buffer.from(headerBytes));
+          blockHeadersHashesFromStream.push(header.hash);
+        });
+
+        allHeadersSettled = blockHeadersHashesGenerated.length >= blocksToGenerate
+          && blockHeadersHashesGenerated
+            .every((hash) => blockHeadersHashesFromStream.includes(hash));
+      }
+
+      if (allHeadersSettled) {
+        const rawChainLock = data.getChainLock();
+        if (rawChainLock) {
+          latestChainLock = new ChainLock(Buffer.from(rawChainLock));
+          stream.destroy();
+          streamEnded = true;
+        }
+      }
+    });
+
+    let streamError;
+    stream.on('error', (e) => {
+      streamError = e;
+    });
+
+    stream.on('end', () => {
+      streamEnded = true;
+    });
+
+    // Generate blocks
+    while (blockHeadersHashesGenerated.length < blocksToGenerate) {
+      const address = new PrivateKey().toAddress(network).toString();
+      const blockHash = (await dapiClient.core.generateToAddress(1, address))[0];
+      const block = new Block(await dapiClient.core.getBlockByHash(blockHash));
+      blockHeadersHashesGenerated.push(block.header.hash);
+      await wait(500);
+    }
```

Comment:
Generate will work only for local network. In case of testnet we already have a lot of blocks

Reply:
Replaced with the transaction broadcasting